### PR TITLE
Add lenses to PParams, create a `[Era]EraPParams` class for each era that needs it

### LINKED ISSUE #2927: Switch `PParams` to use lenses instead of `HasField`
In a recent PR #2901 we've switched all transaction related types to use lenses. Same needs to be done for PParams. 

Also concrete types need to become newtypes around the `*PParamHKD`, rather than the type synonyms. Also renaming fields using proper convention could be done at the same time. 

- [ ] - Use `newtypes` instead of `type` synonyms
- [ ] - Define lenses and provide instances
- [ ] - Rename fields

So the end goal should look something along the lines:

* The type class:
```haskell
class (...) => EraPParams era where
  -- | Protocol parameters
  type PParams era = (r :: Type) | r -> era

  -- | The type of updates to Protocol parameters
  type PParamsUpdate era = (r :: Type) | r -> era

  emptyPParams :: PParams era
  emptyPParamsUpdate:: PParamsUpdate era

  applyPPUpdates :: PParams era -> PParamsUpdate era -> PParams era

  ppMinFeeA_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ppMinfeeB_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ppMaxBBSize_L :: Lens' (PParams (ShelleyEra crypto)) Natural
  ...
  ppA0_L :: Lens' (PParams (ShelleyEra crypto)) NonNegativeInterval
  ppRho_L :: Lens' (PParams (ShelleyEra crypto)) UnitInterval
  ppTau_L :: Lens' (PParams (ShelleyEra crypto))  UnitInterval
  ppD_L :: Lens' (PParams (ShelleyEra crypto))  UnitInterval

  ppuMinfeeA_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ppuMinfeeB_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ppuMaxBBSize_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe Natural)
  ...
  ppuA0_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe NonNegativeInterval)
  ppuRho_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
  ppuTau_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
  ppuD_L :: Lens' (PParamsUpdate (ShelleyEra crypto)) (StrictMaybe UnitInterval)
```

I am not set on the names (for example I am not fond of underscore), but this is the best I could come up with.

Then instances can be defined as:
```haskell
data ShelleyPParamsHKD f era = ShelleyPParams
  { -- | The linear factor for the minimum fee calculation
    sppMinFeeA :: !(HKD f Natural),
  ...
  }

sppMinFeeA_L :: Lens' (ShelleyPParamsHKD f (ShelleyEra crypto)) (HKD f Natural),
sppMinFeeA_L = lens sppMinFeeA (\pp minFeeA -> pp { sppMinFeeA = minFeeA })

newtype ShelleyPParams era = ShelleyPParams (ShelleyPParamsHKD Identity era)

newtype ShelleyPParamsUpdate era = ShelleyPParamsUpdate (ShelleyPParamsHKD StrictMaybe era)

instance CC.Crypto crypto => EraPParams (ShelleyEra crypto) where
  type PParams (ShelleyEra crypto) = ShelleyPParams (ShelleyEra crypto)
  type PParamsUpdate (ShelleyEra crypto) = ShelleyPParamsUpdate (ShelleyEra crypto)

  applyPPUpdates = updatePParams

  ppMinFeeA_L = coerce sppMinFeeA_L
  ...
  ppuMinFeeA_L = coerce sppMinFeeA_L
  ...
```

### PR DESCRIPTION
**Pull request by Soupstraw**
closes #2927

### CODE REVIEW

File: libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+  hkdDL :: Lens' (PParamsHKD f era) (HKD f UnitInterval)
+
+  -- | Extra entropy
+  hkdExtraEntropyL :: Lens' (PParamsHKD f era) (HKD f Nonce)
+
+  -- | Protocol version
+  hkdProtocolVersionL :: Lens' (PParamsHKD f era) (HKD f BT.ProtVer)
+
+  -- | Minimum UTxO value
+  hkdMinUTxOValueL :: Lens' (PParamsHKD f era) (HKD f Coin)
```
Comment (lehins) [inline libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs]: In order to prevent usage of this lens for eras post Mary era this is what we can do:
```suggestion
  hkdMinUTxOValueL :: AtMostEra MaryEra era => Lens' (PParamsHKD f era) (HKD f Coin)
```

This will cause a compilation error when used in Alonzo onwards.

Actual implementation for the function is still necessary, however we can't just use `error`, because ghc will wrongfully  produce a warning. Therefore we need to use a specially crafted function: `Cardano.Ledger.Core.notSupportedInThisEraL`

If `MaryEra` won't work because it is a type synonym we can  use the actual type: `AtMostEra (ShelleyMAEra 'Mary) era`

File: libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+deriving newtype instance
+  (Typeable era, FromCBOR (PParamsHKD Identity era)) => FromCBOR (PParams era)
+
+deriving instance Generic (PParams era)
+
+deriving newtype instance
+  PersistField (PParamsHKD Identity era) => PersistField (PParams era)
+
+deriving newtype instance
+  PersistFieldSql (PParamsHKD Identity era) => PersistFieldSql (PParams era)
```
Comment (JaredCorduan) [inline libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs]: let's orphan this to where it is used

File: eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs [hunk 3]
```diff
@@ -192,6 +193,12 @@ instance Group (MaryValue c) where
 
 instance Abelian (MaryValue c)
 
+instance CC.Crypto c => Split (MaryValue c) where
+  vsplit mv 0 = ([], coin mv)
+  vsplit (MaryValue n ma) m
+    | m <= 0 = error "must split maryvalue into positive parts"
+    | otherwise = (take (fromIntegral m) (repeat (MaryValue (n `div` m) ma)), Coin (n `rem` m))
+
```
Comment (aniketd) [inline eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: @lehins @Soupstraw Need your approval on the correctness of this implementation.

Comment (aniketd) [reply eras/shelley-ma/impl/src/Cardano/Ledger/Mary/Value.hs]: Nevermind. I found an instance declared in `MaryEraGen`

File: libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs (hunk 2 above)
Comment (aniketd) [reply libs/cardano-ledger-core/src/Cardano/Ledger/PParams.hs]: To `Cardano.Ledger.State.Schema` :+1:

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/HardForks.hs [hunk 4]
```diff
@@ -15,74 +14,65 @@ module Cardano.Ledger.Shelley.HardForks
...
   Bool
-aggregatedRewards pp = pvMajor (getField @"_protocolVersion" pp) > 2
+aggregatedRewards pv = pvMajor pv > 2
 
 -- | Starting with protocol version 5, the MIR certs will also be
 -- able to transfer funds between the reserves and the treasury.
 -- Additionally, the semantics for the pervious functionality will
 -- change a bit. Before version 5 redundancies in the instantaneous
--- reward mapping were handled by overriding. Now they are handled
+-- reward mapving were handled by overriding. Now they are handled
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/HardForks.hs:27]: Looks like an accidental change
```suggestion
-- reward mapping were handled by overriding. Now they are handled
```

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/PParams.hs [hunk 5]
```diff
@@ -350,25 +350,27 @@ deriving instance Show (PParams' StrictMaybe era)
...
-              encodeMapElement 1 toCBOR =<< _minfeeB ppup,
-              encodeMapElement 2 toCBOR =<< _maxBBSize ppup,
-              encodeMapElement 3 toCBOR =<< _maxTxSize ppup,
-              encodeMapElement 4 toCBOR =<< _maxBHSize ppup,
-              encodeMapElement 5 toCBOR =<< _keyDeposit ppup,
-              encodeMapElement 6 toCBOR =<< _poolDeposit ppup,
-              encodeMapElement 7 toCBOR =<< _eMax ppup,
-              encodeMapElement 8 toCBOR =<< _nOpt ppup,
-              encodeMapElement 9 toCBOR =<< _a0 ppup,
+            [ encodeMapElement 0  toCBOR =<< _minfeeA ppup,
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/PParams.hs:364]: All of the code is required to be formatted with ormolu, so adding spaces for readability can't be merged

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/Utils.hs [hunk 6]
```diff
@@ -0,0 +1,10 @@
+module Cardano.Ledger.Shelley.Utils 
+  ( Split(..)
+  )
+where
+
+import Cardano.Ledger.Coin (Coin)
+
+class Split v where
```
Comment (lehins) [inline eras/shelley/impl/src/Cardano/Ledger/Shelley/Utils.hs:8]: This class was only used in tests before. No change to PParams should affect that. So, why do you need to move this class into `cardano-ledger-shelley`?

File: eras/shelley/test-suite/bench/Cardano/Ledger/Shelley/Bench/Gen.hs [hunk 7]
```diff
@@ -106,7 +107,7 @@ genBlock ge cs = generate $ GenBlock.genBlock ge cs
 -- 4) get a DPState from the ChainState
 -- 5) get a Transaction (Tx) from GenEnv and ChainState
 
-genTriple ::
+genTriple :: forall era.
```
Comment (lehins) [inline eras/shelley/test-suite/bench/Cardano/Ledger/Shelley/Bench/Gen.hs:110]: This `forall` is not redundant
```suggestion
genTriple ::
```

File: eras/shelley/test-suite/bench/Cardano/Ledger/Shelley/Bench/Gen.hs [hunk 8]
```diff
@@ -1,6 +1,7 @@
 {-# LANGUAGE DataKinds #-}
 {-# LANGUAGE FlexibleContexts #-}
 {-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
```
Comment (lehins) [inline eras/shelley/test-suite/bench/Cardano/Ledger/Shelley/Bench/Gen.hs:4]: This extension is unused
```suggestion
```

File: eras/shelley/test-suite/bench/Main.hs [hunk 9]
```diff
@@ -375,10 +386,19 @@ varyDelegState tag fixed changes initstate action =
...
+  ( EraGen era,
+    Core.PParams era ~ ShelleyPParams era,
+    Mock (EraCrypto era),
+    Embed (Core.EraRule "DELPL" era) (CERTS era),
+    Environment (Core.EraRule "DELPL" era) ~ DelplEnv era,
+    State (Core.EraRule "DELPL" era) ~ DPState (EraCrypto era),
+    Signal (Core.EraRule "DELPL" era) ~ DCert (EraCrypto era),
+    ShelleyTest era
+  ) => IO ()
 -- main=profileValid
```
Comment (lehins) [inline eras/shelley/test-suite/bench/Main.hs:399]: What is all this about?

The only sensible signature for `main` is `main :: IO ()` because you can never provide any argument or type annotation when running a process

File: eras/shelley/test-suite/bench/Main.hs [hunk 10]
```diff
@@ -40,11 +46,15 @@ import Cardano.Ledger.Shelley.LedgerState
     incrementalStakeDistr,
     updateStakeDistribution,
   )
-import Cardano.Ledger.Shelley.PParams (ShelleyPParamsHKD (..))
+-- import Cardano.Ledger.Shelley.PParams (ShelleyPParamsHKD (..))
```
Comment (lehins) [inline eras/shelley/test-suite/bench/Main.hs:49]: Yeah, this import can be removed. It was only imported to bring `HasField` instances into scope.

**Review by lehins — COMMENTED** [review]
I thought I'll skim through the PR. Here are some comments and questions from me

File: eras/shelley/impl/src/Cardano/Ledger/Shelley/PParams.hs (hunk 5 above)
Comment (aniketd) [reply eras/shelley/impl/src/Cardano/Ledger/Shelley/PParams.hs:364]: Yes, noted! I did plan run the `ormolu` script finally, but it makes sense to run it each time something is pushed. :+1:

File: eras/shelley/test-suite/bench/Main.hs (hunk 9 above)
Comment (aniketd) [reply eras/shelley/test-suite/bench/Main.hs:399]: Thanks @lehins! That makes sense. I will remove this.

File: eras/shelley/test-suite/bench/Main.hs (hunk 10 above)
Comment (aniketd) [reply eras/shelley/test-suite/bench/Main.hs:49]: :+1:

File: libs/cardano-protocol-tpraos/cardano-protocol-tpraos.cabal [hunk 11]
```diff
@@ -59,6 +59,7 @@ library
     containers,
     deepseq,
     nothunks,
+    lens,
```
Comment (lehins) [inline libs/cardano-protocol-tpraos/cardano-protocol-tpraos.cabal:62]: Why are we using `lens` instead of `microlens`?

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Core.hs [hunk 12]
```diff
@@ -12,6 +16,12 @@ import Cardano.Ledger.BaseTypes (Network)
 import Cardano.Ledger.Keys (KeyHash (..), KeyRole (..))
 import Cardano.Ledger.SafeHash (SafeHash)
 import Cardano.Ledger.ShelleyMA.Core
+  ( Era (..),
+    EraIndependentScriptIntegrity,
+    EraTxBody (..),
+    ShelleyMAEraTxBody,
+  )
+import qualified Cardano.Ledger.ShelleyMA.Core as X
```
Comment (lehins) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/Core.hs:24]: Why is this change necessary? Is it just to be explicit about all the import used in this module?

**Comment by Soupstraw** [comment]
This PR has become stale

**Comment by lehins** [comment]
@Soupstraw Despite that this PR has diverged a bit from master it is still useful work that will be picked up a bit later. Currently it is blocked by #3171

**Comment by lehins** [comment]
Closing in favor of the work that was built atop of this PR: #3242
